core/modules/ccontrol/testAntlr4GeneratedIR.cc

```diff
@@ -0,0 +1,342 @@
+    Antlr4TestQueries(string const& iQuery, shared_ptr<query::SelectStmt> const& iCompareStmt,
+            string const& iSerializedQuery)
+    : query(iQuery)
+    , compareStmt(iCompareStmt)
+    , serializedQuery(iSerializedQuery)
+    {}
+
+    /// query to test, that will be turned into a SelectStmt by the andlr4-based parser.
+    string query;
+
+    /// comparison query, that will be turned into a SelectStmt by the andlr4-based parser and then that will
+    /// be modified by modFunc
+    shared_ptr<query::SelectStmt> compareStmt;
+
+    /// the query as it should appear after serialization.
+    string serializedQuery;
+};
+
+ostream& operator<<(ostream& os, Antlr4TestQueries const& i) {
+    os << "Antlr4TestQueries(" << i.query << "...)";
+    return os;
+}
+
+
+static const vector<Antlr4TestQueries> ANTLR4_TEST_QUERIES = {
+    // tests NOT LIKE (which is 'NOT LIKE', different than 'NOT' and 'LIKE' operators separately)
+    Antlr4TestQueries(
+        "SELECT sce.filterId, sce.filterName "
+        "FROM Science_Ccd_Exposure AS sce "
+        "WHERE (sce.visit = 887404831) AND (sce.raftName = '3,3') AND (sce.ccdName LIKE '%') "
+        "ORDER BY filterId", // case01/queries/0012.1_raftAndCcd.sql
+        SelectStmt(
+            SelectList(
+                ValueExpr(ValueFactor(ColumnRef("", "sce", "filterId"))),
+                ValueExpr(ValueFactor(ColumnRef("", "sce", "filterName")))
+            ),
+            FromList(TableRef("", "Science_Ccd_Exposure", "sce")),
+            WhereClause(OrTerm(AndTerm(
+                BoolFactor(
+                    PassTerm("("),
+                    BoolTermFactor(OrTerm(AndTerm(BoolFactor(CompPredicate(
+                        ValueExpr(ValueFactor(ColumnRef("", "sce", "visit"))),
+                        SqlSQL2Tokens::EQUALS_OP,
+                        ValueExpr(ValueFactor("887404831"))))))),
+                    PassTerm(")")),
+                BoolFactor(
+                    PassTerm("("),
+                    BoolTermFactor(OrTerm(AndTerm(BoolFactor(CompPredicate(
+                        ValueExpr(ValueFactor(ColumnRef("", "sce", "raftName"))),
+                        SqlSQL2Tokens::EQUALS_OP,
+                        ValueExpr(ValueFactor("'3,3'"))))))),
+                    PassTerm(")")),
+                BoolFactor(
+                    PassTerm("("),
+                    BoolTermFactor(OrTerm(AndTerm(BoolFactor(LikePredicate(
+                        ValueExpr(ValueFactor(ColumnRef("", "sce", "ccdName"))),
+                        ValueExpr(ValueFactor("'%'"))))))),
+                    PassTerm(")"))
+            ))),
+            OrderByClause(OrderByTerm(ValueExpr(ValueFactor(ColumnRef("", "", "filterId")))))
+        ),
+
+        "SELECT sce.filterId,sce.filterName "
+            "FROM Science_Ccd_Exposure AS sce "
+            "WHERE (sce.visit=887404831) AND (sce.raftName='3,3') AND (sce.ccdName LIKE '%') "
+            "ORDER BY filterId"
```

Comment:
A value of the third parameter is the same as the one of the first parameter? Is this always the case? And if so then why this third parameter is needed?